forge/db/controllers/Project.js

```diff
@@ -689,6 +691,56 @@ module.exports = {
         for (const [moduleName, version] of Object.entries(modules)) {
             await app.db.controllers.Project.addProjectModule(instance, moduleName, version)
         }
+    },
+
+    /**
+     * Retrieves the latest state of a specified project.
+     *
+     * @param {string|Number} projectId - The unique identifier of the project whose latest state is to be retrieved.
+     * @returns {string} The latest state of the specified project.
+     */
+    getLatestProjectState: function (app, projectId) {
+        return latestProjectState[projectId]
+    },
+
+    /**
+     * Updates the state of the latest project for a given project ID.
+     *
+     * @function
+     * @param {string|number} projectId - The unique identifier of the project whose state is being updated.
+     * @param {any} state - The new state to be assigned to the project.
+     */
+    setLatestProjectState: function (app, projectId, state) {
+        latestProjectState[projectId] = state
+    },
+
+    /**
+     * Clears the latest project state for a specific project.
+     *
+     * This function deletes the entry in the `latestProjectState` object corresponding
+     * to the provided project's ID. The state deletion is context-specific to the project
+     * and associated application.
+     *
+     * @param {String|Number} projectId - The project id whose latest state should be cleared.
+     */
+    clearLatestProjectState: function (app, projectId) {
+        delete latestProjectState[projectId]
+    },
+
+    /**
+     * Updates the latest project state for the specified app and project based on the given state.
+     * If the state is 'running' or 'stopped', it clears the latest driver state.
+     * Otherwise, it sets the latest driver state to the new value.
+     *
+     * @param {String|Number} projectId - The project id related to the driver state update.
+     * @param {string} state - The new state to update, can be 'running', 'stopped', or other valid states.
+     */
+    updateLatestProjectState: function (app, projectId, state) {
+        if (['running', 'stopped'].includes(state)) {
```

Comment:
in hindsight, this may not be the best way to clear the cached state. A better alternative would be to clear the state only when the desired state of the instance (stored in the db) matches the state we're trying to update to.

Otherwise we might get into situations where ok, we received a stopped state, we'll remove it from the cache but the desired state of the instance is running.